p2p/src/main/java/bisq/network/p2p/network/TorNetworkNode.java

```diff
@@ -288,13 +288,13 @@ public void run() {
                 log.error("Tor node creation failed: " + (e.getCause() != null ? e.getCause().toString() : e.toString()));
                 restartTor(e.getMessage());
             } catch (IOException e) {
-                log.error("Could not connect to running Tor: "
-                        + e.getMessage());
-
-                // Seems a bit harsh, but since we cannot connect to Tor, we cannot do nothing.
+                log.error("Could not connect to running Tor: " + e.getMessage());
+                // Since we cannot connect to Tor, we cannot do nothing.
                 // Furthermore, we have no hidden services started yet, so there is no graceful
                 // shutdown needed either
-                System.exit(1);
+                UserThread.execute(() -> {
+                    setupListeners.stream().forEach(s -> s.onSetupFailed(new RuntimeException(e.getMessage())));
+                });
             } catch (Throwable ignore) {
             }
```

Comment:
@freimair Since I removed `restartTor(throwable.getMessage());` from onFailure below since I didn't want it to be called after handling IOException above, perhaps I should add it to this ignored exception?

Reply:
Maybe we can force `onSuccess` by `return null;` after triggering the `onSetupFailed` and let the `onFailure` be?

And yes, this class cries for being refactored :)

Reply:
In fact we are already doing that. Let me review this again. I should be able to add restartTor back to onFailure.